security/src/main/java/io/micronaut/security/x509/X509AuthenticationFetcher.java

```diff
@@ -0,0 +1,121 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.security.x509;
+
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.http.HttpRequest;
+import io.micronaut.security.authentication.Authentication;
+import io.micronaut.security.filters.AuthenticationFetcher;
+import io.micronaut.security.token.TokenAuthenticationFetcher;
+import jakarta.inject.Singleton;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static io.micronaut.core.util.StringUtils.TRUE;
+import static java.util.regex.Pattern.CASE_INSENSITIVE;
+
+/**
+ * Creates an Authentication if an X.509 client certificate is present and a
+ * name (CN) can be extracted.
+ *
+ * @author Burt Beckwith
+ * @since 3.2
+ */
+@Singleton
+@Requires(property = X509ConfigurationProperties.PREFIX + ".enabled", value = TRUE)
+public class X509AuthenticationFetcher implements AuthenticationFetcher {
+
+    /**
+     * The order of the fetcher.
+     */
+    public static final int ORDER = TokenAuthenticationFetcher.ORDER - 200;
+
+    private final Pattern subjectDnPattern;
+
+    public X509AuthenticationFetcher(X509Configuration x509Configuration) {
+        subjectDnPattern = Pattern.compile(x509Configuration.getSubjectDnRegex(), CASE_INSENSITIVE);
+    }
+
+    @Override
+    public int getOrder() {
+        return ORDER;
+    }
+
+    @Override
+    public Publisher<Authentication> fetchAuthentication(HttpRequest<?> request) {
+        return Mono.create(emitter -> {
+            Optional<Authentication> authentication = createAuthentication(request);
+            emitter.success(authentication.orElse(null));
+        });
+    }
+
+    /**
+     * Creates an {@link X509Authentication} from information in an {@link X509Certificate}
+     * if one is present in the request.
+     *
+     * @param request the request
+     * @return the authentication if the certificate exists and contains a valid name
+     */
+    protected Optional<Authentication> createAuthentication(HttpRequest<?> request) {
+        Optional<Certificate> optionalCertificate = request.getCertificate();
+        if (optionalCertificate.isPresent()) {
+            Certificate certificate = optionalCertificate.get();
+            if (certificate instanceof X509Certificate) {
```

Comment:
This should support subclasses of `X509Certificate`
```java
X509Certificate.class.isAssignableFrom(certificate.getClass())
```

Reply:
A subclass of X509 is an instance of X509